File: python/295.py
import heapq


class MedianFinder:
    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.less = []  # max-heap
        self.ln = 0
        self.more = []  # min-heap
        self.mn = 0

    def addNum(self, num):
        """
        Adds a num into the data structure.
        :type num: int
        :rtype: void
        """
        if not self.more or num > self.more[0]:
            heapq.heappush(self.more, num)
            self.mn += 1
        else:
            heapq.heappush(self.less, -num)
            self.ln += 1
        if self.mn - self.ln > 1:
            heapq.heappush(self.less, -heapq.heappop(self.more))
            self.mn -= 1
            self.ln += 1
        elif self.ln - self.mn > 1:
            heapq.heappush(self.more, -heapq.heappop(self.less))
            self.mn += 1
            self.ln -= 1
        assert abs(self.mn - self.ln) <= 1, (self.less, self.more)

    def findMedian(self):
        """
        Returns the median of current data stream
        :rtype: float
        """
        if self.ln == self.mn:
            return (-self.less[0] + self.more[0]) / 2.
        elif self.ln > self.mn:
            return -self.less[0]
        else:
            return self.more[0]


# Your MedianFinder object will be instantiated and called as such:
mf = MedianFinder()
mf.addNum(-1)
print(mf.findMedian())
mf.addNum(-2)
print(mf.findMedian())
mf.addNum(-3)
print(mf.findMedian())
mf.addNum(-4)
print(mf.findMedian())
mf.addNum(-5)
print(mf.findMedian())

File: cpp/295.cpp
#include <queue>
#include <vector>
#include <iostream>
using namespace std;

class MedianFinder {
  priority_queue<int, std::vector<int>, greater<int>> more;
  priority_queue<int> less;
  int mn=0, ln=0;
public:
  void addNum(int num) {
    if(more.size() == 0 || num > more.top())
      more.push(num);
    else
      less.push(num);
    if(int(more.size()) - int(less.size()) > 1){
      less.push(more.top()); more.pop();
    }else if(int(less.size()) - int(more.size()) > 1){
      more.push(less.top()); less.pop();
    }
  }

  double findMedian() {
    if(more.size() == less.size())
      return (more.top() + less.top()) / 2.0;
    else if(more.size() > less.size())
      return more.top();
    else
      return less.top();
  }
};

int main(){
  auto mf = MedianFinder();
  mf.addNum(1);
  mf.addNum(2);
  mf.addNum(3);
  cout << mf.findMedian() << endl;
}
